File: frontend/src/infrastructure/mappers/productHistoryMapper.ts
import ProductHistory from "../../domain/models/IProductHistory";
import IProductHistoryApiModel from "../apiModels/IProductHistoryApiModel";

const productHistoryMapper = {
    apiToDomain: (source: IProductHistoryApiModel): ProductHistory => {
        return new ProductHistory({
            id: source.id,
            name: source.name,
            images: source.images,
            description: source.description,
            price: source.price,
            productId: source.productId,
            validFrom: new Date(source.validFrom),
            validTo: source.validTo == null ? null : new Date(source.validTo),
        });
    },
};

export default productHistoryMapper;

File: Infrastructure/Mappers/ProductHistoryMapper.cs
using Domain.Models;
using Domain.ValueObjects.Product;
using Domain.ValueObjects.ProductHistory;
using Domain.ValueObjects.Shared;
using Infrastructure.DbEntities;

namespace Infrastructure.Mappers;

public static class ProductHistoryMapper
{
    public static ProductHistory ToDomain(ProductHistoryDbEntity source)
    {
        return new ProductHistory(
            id: ProductHistoryId.ExecuteCreate(source.Id),
            name: source.Name,
            images: source.Images,
            price: Money.ExecuteCreate(source.Price),
            productId: ProductId.ExecuteCreate(source.OriginalProductId),
            validityRange: ProductHistoryValidityRange.ExecuteCreate(
                validFrom: source.ValidFrom,
                validTo: source.ValidTo
            ),
            description: source.Description
        );
    }

    public static ProductHistoryDbEntity ToDbModel(ProductHistory source)
    {
        return new ProductHistoryDbEntity(
            id: source.Id.Value,
            name: source.Name,
            images: source.Images,
            description: source.Description,
            price: source.Price.Value,
            productId: source.ProductId.Value,
            originalProductId: source.ProductId.Value,
            validFrom: source.ValidityRange.ValidFrom,
            validTo: source.ValidityRange.ValidTo
        );
    }
}